### PR TITLE
fix: spy on popup apis like alert

### LINKED ISSUE #3061: bug (browser-runner): alert stales test execution
### Describe the bug

Heyho 👋 

since I discovered this edge case in WebdriverIO I thought I would report it in Vitest as well. If `alert` is called, all JS execution is stopped including the communication with the Vite server.

I am thinking about the following options:
- throw an error and advise users to mock this method
- auto mock it as part of the test setup

What do you think? 

### Reproduction

Given the following test:

```ts
import { test } from 'vitest'

test('haha', () => {
    alert("UOOO")
    console.log('Hello World')
})
```

The console command is never executed.

### System Info

```shell
System:
    OS: macOS 12.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 711.84 MB / 32.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.18.1 - ~/.nvm/versions/node/v16.18.1/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.18.1/bin/yarn
    npm: 8.19.2 - ~/.nvm/versions/node/v16.18.1/bin/npm
  Browsers:
    Brave Browser: 111.1.49.120
    Chrome: 111.0.5563.64
    Firefox: 109.0
    Firefox Nightly: 111.0a1
    Safari: 15.2
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by Aslemammad**
Resolves #3061 

We try to spy the `alert`, `confirm` and `prompt` so we do not let the browser stop and not send any event to vitest. We return null and show a warning to the user, so it'd be better if the user mock them.

### CODE REVIEW

File: packages/browser/src/client/popup.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+${name}(${formatedParams})
+expect(${name}).toHaveBeenCalled()
+\`\`\``)
+  return null
+}
+
+export const setupPopupSpy = () => {
+  globalThis.alert = showPopupWarning('alert')
+  // @ts-expect-error mocking native popup apis
+  globalThis.confirm = showPopupWarning('confirm')
```
Comment (sheremet-va) [inline packages/browser/src/client/popup.ts:20]: `confirm` always returns `boolean`, so you should update the return value here instead of suppressing typescript error

File: packages/browser/src/client/popup.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+const showPopupWarning = (name: string) => (...params: any[]) => {
+  const formatedParams = params.map(p => JSON.stringify(p)).join(', ')
+
+  console.warn(`Vitest encountered a \`${name}\` call with parameters: ${
+      formatedParams}. Vitest cannot handle \`${
+      name}\` by default in the browser, so it returned null. If you need to handle this case, you can mock the \`${
+      name}\` call yourself like this:
+
+\`\`\`
+vi.spyOn(window, '${name}').mockReturnValue('your value')
```
Comment (sheremet-va) [inline packages/browser/src/client/popup.ts:10]: The text was just an example. We also need to import `vi` from `vitest` here to show where it comes from to the new users.

I also don't think using "`" does anything here for the preview. Do you have pictures how it looks like in browser and in the terminal? Can we improve the look somehow?

**Review by sheremet-va — CHANGES_REQUESTED** [review]
I also assumed we will have a page in documentation for this types of errors that we can also reference to? Or at least have a link to mocking section?

File: packages/browser/src/client/popup.ts (hunk 2 above)
Comment (Aslemammad) [reply packages/browser/src/client/popup.ts:10]: image

File: packages/browser/src/client/popup.ts (hunk 1 above)
Comment (Aslemammad) [reply packages/browser/src/client/popup.ts:20]: Yea, tried that first, but thought returning false may be wrong. Let me change that, I think it's better.

File: packages/browser/src/client/popup.ts (hunk 2 above)
Comment (Aslemammad) [reply packages/browser/src/client/popup.ts:10]: Yes, I'll update it. but around the `, I think they separate the text from the code properly.

image

Comment (Aslemammad) [reply packages/browser/src/client/popup.ts:10]: Here's the new version, I also added some docs.

image

Comment (Aslemammad) [reply packages/browser/src/client/popup.ts:10]: image

File: packages/browser/src/client/dialog.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+const showPopupWarning = <T>(name: string, value: T, defaultValue?: T) => (...params: any[]) => {
+  const formatedParams = params.map(p => JSON.stringify(p)).join(', ')
+
+  console.warn(`Vitest encountered a \`${name}(${formatedParams})\` call that it cannot handle by default, so it returned \`${value}\`. Read more in https://vitest.dev/guide/browser#thread-blocking-dialogs.
```
Comment (Aslemammad) [inline packages/browser/src/client/dialog.ts:4]: @sheremet-va You also need to update the tests for the new text! I can do it but I'm busy with the playwright integration right now.

**Comment by charlieforward9** [comment]
Any chance we can get `window.print` added here quickly? I am waiting on a [`vi.mock` bug](https://github.com/vitest-dev/vitest/issues/7290) to get fixed and could use this default implementation for the time being.
